pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerSemaphoreTest.java

```diff
@@ -0,0 +1,135 @@
+            futures.add(producer.newMessage().value(("Semaphore-test-" + i).getBytes()).sendAsync());
+        }
+
+        FutureUtil.waitForAll(futures).get();
+        Assert.assertEquals(producer.getSemaphore().availablePermits(), pendingQueueSize);
+        futures.clear();
+
+        // Simulate replicator, non batching message but `numMessagesInBatch` of message metadata > 1
+        for (int i = 0; i < messages / 2; i++) {
+            PulsarApi.MessageMetadata.Builder builder = PulsarApi.MessageMetadata.newBuilder();
+            builder.setNumMessagesInBatch(10);
+            MessageImpl<byte[]> msg = MessageImpl.create(builder, ByteBuffer.wrap(new byte[0]), Schema.BYTES);
+            futures.add(producer.sendAsync(msg));
+        }
+        FutureUtil.waitForAll(futures).get();
+
+        // Here must ensure that the semaphore available permits is 0
+        Assert.assertEquals(producer.getSemaphore().availablePermits(), pendingQueueSize);
+
+        // Acquire 5 and not wait the send ack call back
+        for (int i = 0; i < messages / 2; i++) {
+            producer.newMessage().value(("Semaphore-test-" + i).getBytes()).sendAsync();
+        }
+
+        // Here must ensure that the Semaphore a acquired 5
+        Assert.assertEquals(producer.getSemaphore().availablePermits(), pendingQueueSize - messages / 2);
+
+    }
+
+    /**
+     * We use semaphore to limit the pending send, so we must ensure that the thread of sending message never block
+     * at the pending message queue. If not, the dead lock might occur. Here is the related issue to describe the
+     * dead lock happens {https://github.com/apache/pulsar/issues/5585}
+     */
+    @Test(timeOut = 30000)
+    public void testEnsureNotBlockOnThePendingQueue() throws PulsarClientException {
+
+        final int pendingQueueSize = 10;
+
+        @Cleanup
+        ProducerImpl<byte[]> producer = (ProducerImpl<byte[]>) pulsarClient.newProducer()
+                .topic("testProducerSemaphoreAcquire")
+                .maxPendingMessages(pendingQueueSize)
+                .enableBatching(false)
+                .blockIfQueueFull(true)
+                .create();
+
+        final int messages = 20;
+        final List<CompletableFuture<MessageId>> futures = new ArrayList<>(messages);
+
+        // Simulate replicator, non batching message but `numMessagesInBatch` of message metadata > 1
+        for (int i = 0; i < messages; i++) {
+            PulsarApi.MessageMetadata.Builder builder = PulsarApi.MessageMetadata.newBuilder();
+            builder.setNumMessagesInBatch(10);
+            MessageImpl<byte[]> msg = MessageImpl.create(builder, ByteBuffer.wrap(new byte[0]), Schema.BYTES);
+            futures.add(producer.sendAsync(msg));
+        }
+
+        FutureUtil.waitForAll(futures);
+        futures.clear();
+
+        for (int i = 0; i < messages; i++) {
+            futures.add(producer.newMessage().value(("Semaphore-test-" + i).getBytes()).sendAsync());
+        }
```

Comment:
@weishuisheng What are you trying to test here? At this point, the availablePermits can be anything between 0 and 10, depending on how fast the broker has responded. Try putting a Thread.sleep(1000) here and see that it fails every time.  My understanding of the issue is that you want to ensure that batches take N permits and release N permits when finished, correct?

Reply:
@ivankelly I added the unit test for this PR, I will take a look.